jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java

```diff
@@ -376,7 +376,7 @@ public long getContentLengthValue()
         @Override
         public ByteBuffer getByteBuffer()
         {
-            return _buffer;
+            return _buffer == null ? null : _buffer.asReadOnlyBuffer();
```

Comment:
I don't think this should be done here, it should be done in the constructor as `_buffer` is final.

Reply:
Have a look at the test I added. The bug is that the buffer returned by this method gets consumed by the write(). So someone has to duplicate the buffer: either here, or just before write(), where I added an extra comment.

I decided here arbitrarily to do it in `getBuffer()`, which is my @gregw pointed out that maybe that method shouldn't be called getBuffer but something like newBuffer or createBuffer. And that this bug happened because this contract isn't documented.

Reply:
I remain 99% convinced that the whole `HttpContent.getByteBuffer` approach is fundamentally flawed.  The implementations need to know why they are being called before they can know what type of buffer they should return.  If the `HttpContent` is being cached, then the buffer needs to be duplicated, but if it is not being cached, then a duplication is wasted effort on a buffer that is being thrown away anyway.

This bug/PR has only increased that conviction.   This is something we really need to resolve early in the new year and I think we need to reconsider @lachlan-roberts work on this.

But for now, we need a fix so we can serve content more than once.     Either the buffer needs to be duplicated/sliced within getByteBuffer implementations; or the caller needs to duplicate/slice before it does anything to modify the psoition/limit.  Pick one approach, document it, implement it.   Get this working and then we review.